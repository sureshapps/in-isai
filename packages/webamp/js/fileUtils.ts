import invariant from "invariant";
import { IMetadataApi } from "./types";
import { IAudioMetadata } from "music-metadata-browser"; // Import music-metadata type definitions
import * as Utils from "./utils";

type MediaDataType = string | ArrayBuffer | Blob;

export async function genMediaTags(
  file: MediaDataType,
  musicMetadata: IMetadataApi
): Promise<IAudioMetadata> {
  invariant(
    file != null,
    "Attempted to get the tags of media file without passing a file"
  );

  const options = {
    duration: true,
    skipPostHeaders: true, // avoid unnecessary data to be read
  };

  if (typeof file === "string") {
    if (
      "parseWebStream" in musicMetadata &&
      typeof musicMetadata.parseWebStream === "function"
    ) {
      const response = await fetch(file);
      if (!response.ok) {
        throw new Error(
          `Failed to fetch URL: ${file}, status: ${response.status}`
        );
      }

      // https://github.com/Borewit/music-metadata/issues/2455
      // There's currently an issue where some URLs will fail to parse id3 tags
      // when using parseWebStream. This approach can work around it. However,
      // My current assumption is that this is an issue mostly specific to that
      // individual file and not a wide spread issue, but if we find it happens
      // more broadly we can deopt to using parseBlob as below.

      // const blob = await response.blob();
      // return musicMetadata.parseBlob(blob, options);

      const webStream = response.body;
      if (webStream == null) {
        throw new Error("Response body is null, cannot parse metadata.");
      }
      return musicMetadata.parseWebStream(webStream, undefined, options);
    }
    if (
      "fetchFromUrl" in musicMetadata &&
      typeof musicMetadata.fetchFromUrl === "function"
    ) {
      return musicMetadata.fetchFromUrl(file, options);
    }
    throw new Error("No suitable method available to parse URL");
  }
  // Assume Blob
  return musicMetadata.parseBlob(file as Blob, options);
}

export function genMediaDuration(url: string): Promise<number> {
  invariant(
    typeof url === "string",
    "Attempted to get the duration of media file without passing a url"
  );
  return new Promise((resolve, reject) => {
    // TODO: Does this actually stop downloading the file once it's
    // got the duration?
    const audio = document.createElement("audio");
    audio.crossOrigin = "anonymous";

    const durationChange = () => {
      resolve(audio.duration);
      audio.removeEventListener("durationchange", durationChange);
      audio.removeEventListener("error", errorHandler);
      audio.src = "";
      // TODO: Not sure if this really gets cleaned up.
    };

    const errorHandler = (e: Event) => {
      audio.removeEventListener("durationchange", durationChange);
      audio.removeEventListener("error", errorHandler);
      reject(e);
    };

    audio.addEventListener("durationchange", durationChange);
    audio.addEventListener("error", errorHandler);
    audio.src = url;
  });
}

export async function genArrayBufferFromFileReference(
  fileReference: File
): Promise<any> {
  invariant(
    fileReference != null,
    "Attempt to get an ArrayBuffer without assigning a fileReference"
  );
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = () => {
      resolve(reader.result as ArrayBuffer);
    };
    reader.onerror = reject;
    reader.readAsArrayBuffer(fileReference);
  });
}

export async function genStringFromFileReference(
  fileReference: File
): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = () => {
      resolve(reader.result as string);
    };
    reader.onerror = reject;
    reader.readAsText(fileReference);
  });
}

interface PromptForFileReferenceOptions {
  accept?: string | null;
  directory?: boolean;
}

export async function promptForFileReferences(
  { accept, directory = false }: PromptForFileReferenceOptions = {
    accept: null,
    directory: false,
  }
): Promise<FileList> {
  return new Promise<FileList>((resolve) => {
    // Does this represent a memory leak somehow?
    // Can this fail? Do we ever reject?
    const fileInput = document.createElement("input");
    if (accept) fileInput.setAttribute("accept", accept);
    fileInput.type = "file";
    fileInput.multiple = true;
    // @ts-ignore Non-standard
    fileInput.webkitdirectory = directory;
    // @ts-ignore Non-standard
    fileInput.directory = directory;
    // @ts-ignore Non-standard
    fileInput.mozdirectory = directory;
    // Not entirely sure why this is needed, since the input
    // was just created, but somehow this helps prevent change
    // events from getting swallowed.
    // https://stackoverflow.com/a/12102992/1263117

    // @ts-ignore Technically you can't set this to null, it has to be a string.
    // But I don't feel like retesting it, so I'll leave it as null
    fileInput.value = null;
    fileInput.addEventListener("change", (e: Event) => {
      const files = (<HTMLInputElement>e.target).files;
      resolve(files as FileList);
    });
    fileInput.click();
  });
}

function urlIsBlobUrl(url: string): boolean {
  return /^blob:/.test(url);
}

export function curUrlFromByteArray(arr: Uint8Array) {
  const base64 = Utils.base64FromDataArray(arr);
  return `data:image/x-win-bitmap;base64,${base64}`;
}

// This is not perfect, but... meh: https://stackoverflow.com/a/36756650/1263117
export function filenameFromUrl(url: string): string | null {
  if (urlIsBlobUrl(url)) {
    return null;
  }

  const lastSegment = url.split("/").pop();
  if (lastSegment == null) {
    return null;
  }
  return lastSegment.split("#")[0].split("?")[0];
}
