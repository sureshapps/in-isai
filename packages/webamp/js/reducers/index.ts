import { AppState, Action } from "../types";
import { Reducer, combineReducers } from "redux";

import playlist from "./playlist";
import windows from "./windows";
import media from "./media";
import display from "./display";
import userInput from "./userInput";
import equalizer from "./equalizer";
import network from "./network";
import settings from "./settings";
import tracks from "./tracks";
import milkdrop from "./milkdrop";

const reducer: Reducer<AppState, Action, never> = combineReducers({
  userInput,
  windows,
  display,
  settings,
  equalizer,
  playlist,
  media,
  network,
  tracks,
  milkdrop,
});

export default reducer;
