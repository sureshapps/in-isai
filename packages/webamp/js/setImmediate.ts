// A Fork of the NPM module `setimmediate`
// I've adapted this to only include the browser implementation and to not
// mutate the global object and instead export the `setImmediate` function.

type Handle = { callback: Function; args: any[] };
type TaskByHandle = { [handle: number]: Handle };

let nextHandle = 1; // Spec says greater than zero
const tasksByHandle: TaskByHandle = {};
let currentlyRunningATask = false;
let registerImmediate: (arg0: number) => void;

export function setImmediate(callback: Function) {
  // Copy function arguments
  const args = new Array(arguments.length - 1);
  for (let i = 0; i < args.length; i++) {
    args[i] = arguments[i + 1];
  }
  // Store and register the task
  const task = { callback: callback, args: args };
  tasksByHandle[nextHandle] = task;
  registerImmediate(nextHandle);
  return nextHandle++;
}

function clearImmediate(handle: number) {
  delete tasksByHandle[handle];
}

function run(task: Handle) {
  const callback = task.callback;
  const args = task.args;
  switch (args.length) {
    case 0:
      callback();
      break;
    case 1:
      callback(args[0]);
      break;
    case 2:
      callback(args[0], args[1]);
      break;
    case 3:
      callback(args[0], args[1], args[2]);
      break;
    default:
      // eslint-disable-next-line prefer-spread
      callback.apply(undefined, args);
      break;
  }
}

function runIfPresent(handle: number) {
  // From the spec: "Wait until any invocations of this algorithm started before this one have completed."
  // So if we're currently running a task, we'll need to delay this invocation.
  if (currentlyRunningATask) {
    // Delay by doing a setTimeout. setImmediate was tried instead, but in Firefox 7 it generated a
    // "too much recursion" error.
    setTimeout(runIfPresent, 0, handle);
  } else {
    const task = tasksByHandle[handle];
    if (task) {
      currentlyRunningATask = true;
      try {
        run(task);
      } finally {
        clearImmediate(handle);
        currentlyRunningATask = false;
      }
    }
  }
}

function canUsePostMessage() {
  // The test against `importScripts` prevents this implementation from being installed inside a web worker,
  // where `global.postMessage` means something completely different and can't be used for this purpose.
  if (!window.importScripts) {
    let postMessageIsAsynchronous = true;
    const oldOnMessage = window.onmessage;
    window.onmessage = function () {
      postMessageIsAsynchronous = false;
    };
    window.postMessage("", "*");
    window.onmessage = oldOnMessage;
    return postMessageIsAsynchronous;
  }
}

function installPostMessageImplementation() {
  // Installs an event handler on `global` for the `message` event: see
  // * https://developer.mozilla.org/en/DOM/window.postMessage
  // * http://www.whatwg.org/specs/web-apps/current-work/multipage/comms.html#crossDocumentMessages

  const messagePrefix = `setImmediate$${Math.random()}$`;

  window.addEventListener(
    "message",
    (event) => {
      if (
        event.source === window &&
        typeof event.data === "string" &&
        event.data.indexOf(messagePrefix) === 0
      ) {
        runIfPresent(+event.data.slice(messagePrefix.length));
      }
    },
    false
  );

  registerImmediate = function (handle) {
    window.postMessage(messagePrefix + handle, "*");
  };
}

function installSetTimeoutImplementation() {
  registerImmediate = function (handle) {
    setTimeout(runIfPresent, 0, handle);
  };
}

if (canUsePostMessage()) {
  // For non-IE10 modern browsers
  installPostMessageImplementation();
} else {
  // For older browsers
  installSetTimeoutImplementation();
}
