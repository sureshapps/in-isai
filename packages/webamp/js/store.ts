import {
  legacy_createStore as createStore,
  applyMiddleware,
  Middleware as ReduxMiddleware,
} from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "@redux-devtools/extension";
import reducer from "./reducers";
import mediaMiddleware from "./mediaMiddleware";
import { merge } from "./utils";
import { IMedia } from "./media";
import Emitter from "./emitter";
import {
  Extras,
  Dispatch,
  Action,
  PartialState,
  Middleware,
  Store,
} from "./types";

const compose = composeWithDevTools({
  actionsDenylist: ["UPDATE_TIME_ELAPSED", "STEP_MARQUEE"],
});

export default function createWebampStore(
  media: IMedia,
  actionEmitter: Emitter,
  customMiddlewares: Middleware[] = [],
  stateOverrides: PartialState | undefined,
  extras: Extras
): Store {
  let initialState;
  if (stateOverrides) {
    initialState = merge(
      reducer(undefined, { type: "@@init" }),
      stateOverrides
    );
  }

  const emitterMiddleware = () => (next: Dispatch) => (action: Action) => {
    actionEmitter.trigger(action.type, action);
    return next(action);
  };

  const middlewares: Middleware[] = [
    thunk.withExtraArgument(extras),
    mediaMiddleware(media),
    emitterMiddleware,
    ...customMiddlewares,
  ];

  // @ts-expect-error Typing of these is too hard to get right, so we cheat
  const coercedMiddlewares: ReduxMiddleware[] = middlewares;

  const enhancer = compose(applyMiddleware(...coercedMiddlewares));

  return createStore(reducer, initialState, enhancer);
}
