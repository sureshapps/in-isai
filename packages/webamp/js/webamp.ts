import JSZip from "jszip";
import * as musicMetadata from "music-metadata";
import { IMetadataApi, Options } from "./types";
import WebampLazy, { PrivateOptions } from "./webampLazy";

// There is some bug between how JSZip pulls in setimmediate (which it expects
// to polyfill `window.setimmediate` and our bundler set. The result is that one
// of our bundles is missing the polyfill. If we call JSZip code from within
// that bundle the polyfill is not present and we get an error.
//
// To work around this we manually import the helper function which is the one
// place JSZip actually uses `setImmediate` and then set it to use our fork of
// the polyfill which does not require adding a property to `window`.
import { setImmediate } from "./setImmediate";
// @ts-ignore
import Utils from "jszip/lib/utils";
// @ts-ignore
Utils.delay = function (callback, args, self) {
  setImmediate(() => {
    callback.apply(self || null, args || []);
  });
};

export type * from "./types";

export default class Webamp extends WebampLazy {
  constructor(options: Options & PrivateOptions) {
    super({
      ...options,
      requireJSZip: async () => JSZip,
      requireMusicMetadata: async (): Promise<IMetadataApi> => {
        // @ts-ignore
        return musicMetadata;
      },
    });
  }
}

// Bit of a hack here. This overwrites the value set in WebampLazy.ts
// @ts-ignore
window.Webamp = Webamp;
