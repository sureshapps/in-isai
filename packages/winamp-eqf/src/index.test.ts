import { join } from "path";
import { readFileSync } from "fs";
import { parser, creator } from "./index.js";
import bufferToArrayBuffer from "buffer-to-arraybuffer";

// TODO: Abstract this into its own library.
declare global {
  namespace jest {
    interface Matchers<R> {
      arrayBufferToEqual(expected: ArrayBuffer): R;
    }
  }
}

expect.extend({
  arrayBufferToEqual(received: ArrayBuffer, argument: ArrayBuffer) {
    if (received.byteLength !== argument.byteLength) {
      return {
        message: () =>
          `ArrayBuffers do not match. Expected length ${received.byteLength} but got ${argument.byteLength}`,
        pass: false,
      };
    }
    const a = new Uint8Array(received);
    const b = new Uint8Array(argument);
    for (let i = 0; i < a.length; i++) {
      if (a[i] !== b[i]) {
        return {
          message: () =>
            `ArrayBuffers do not match. Expected ${a[i]} to equal ${b[i]} at index ${i}`,
          pass: false,
        };
      }
    }
    return {
      message: () => `ArrayBuffers are equal.`,
      pass: true,
    };
  },
});

const fixtures = [
  // All bands max, preamp mid
  "max.EQF",
  // All bands min, preamp mid
  "min.EQF",
  // All bands mid, preamp mid
  "midline.EQF",
  // All bands mid, preamp max
  "preampMax.EQF",
  // All bands mid, preamp min
  "preampMin.EQF",
  "random.EQF",
  "winamp_sample.q1",
  "winamp.q1",
];

describe("parser", () => {
  fixtures.forEach((fileName) => {
    const buffer = readFileSync(join(__dirname, "../sample_data", fileName));
    const arrayBuffer = bufferToArrayBuffer(buffer);
    test(`${fileName}`, () => {
      expect(() => parser(arrayBuffer)).not.toThrow();
    });
  });

  test("parsed file snapshot", () => {
    const buffer = readFileSync(join(__dirname, "../sample_data/random.EQF"));
    const arrayBuffer = bufferToArrayBuffer(buffer);
    expect(parser(arrayBuffer)).toMatchSnapshot();
  });
});

describe("creator", () => {
  fixtures.forEach((fileName) => {
    const buffer = readFileSync(join(__dirname, "../sample_data", fileName));
    const arrayBuffer = bufferToArrayBuffer(buffer);
    const data = parser(arrayBuffer);
    test(`functional round trip: ${fileName}`, () => {
      // Test that parsing the created data returns the same logical data
      expect(parser(creator(data))).toEqual(data);
    });
  });
});

// Only test byte-for-byte equality on .EQF files, not .q1 files
const eqfFixtures = [
  "max.EQF",
  "min.EQF",
  "midline.EQF",
  "preampMax.EQF",
  "preampMin.EQF",
  "random.EQF",
];

describe("creator byte-for-byte", () => {
  eqfFixtures.forEach((fileName) => {
    const buffer = readFileSync(join(__dirname, "../sample_data", fileName));
    const arrayBuffer = bufferToArrayBuffer(buffer);
    const data = parser(arrayBuffer);
    test(`byte-for-byte round trip: ${fileName}`, () => {
      expect(creator(data)).arrayBufferToEqual(arrayBuffer);
    });
  });
});

describe("integration", () => {
  test("can parse and recreate Winamp's sample file", () => {
    const buffer = readFileSync(
      join(__dirname, "../sample_data/winamp_sample.q1")
    );
    const arrayBuffer = bufferToArrayBuffer(buffer);
    const data = parser(arrayBuffer);
    expect(data.presets).toHaveLength(4);
    expect(data.presets[0].name).toEqual("Normal");
    expect(data.presets[1].name).toEqual("Clear");
    expect(data.presets[2].name).toEqual("Alex");
    expect(data.presets[3].name).toEqual("Tare");

    // Functional round-trip test for .q1 files (not byte-for-byte)
    const recreated = creator(data);
    expect(parser(recreated)).toEqual(data);
  });
});
