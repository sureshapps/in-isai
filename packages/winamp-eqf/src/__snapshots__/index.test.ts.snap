// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser parsed file snapshot 1`] = `
{
  "presets": [
    {
      "hz1000": 33,
      "hz12000": 15,
      "hz14000": 47,
      "hz16000": 33,
      "hz170": 35,
      "hz3000": 48,
      "hz310": 1,
      "hz60": 64,
      "hz600": 33,
      "hz6000": 33,
      "name": "Entry1",
      "preamp": 55,
    },
  ],
  "type": "Winamp EQ library file v1.1",
}
`;
